sycl/plugins/unified_runtime/ur/ur.hpp

```diff
@@ -99,6 +99,101 @@ class SpinLock {
+//
+template <class T> struct ZeCache : private T {
+  // The initialization function takes a reference to the data
+  // it is going to initialize, since it is private here in
+  // order to disallow access other than through "->".
+  //
+  using InitFunctionType = std::function<void(T &)>;
+  InitFunctionType Compute{nullptr};
+  bool Computed{false};
+  pi_mutex ZeCacheMutex;
+
+  ZeCache() : T{} {}
+
+  // Access to the fields of the original T data structure.
+  T *operator->() {
+    std::unique_lock<pi_mutex> Lock(ZeCacheMutex);
+    if (!Computed) {
+      Compute(*this);
+      Computed = true;
+    }
+    return this;
+  }
+};
+
+// This wrapper around std::atomic is created to limit operations with reference
+// counter and to make allowed operations more transparent in terms of
+// thread-safety in the plugin. increment() and load() operations do not need a
+// mutex guard around them since the underlying data is already atomic.
+// decrementAndTest() method is used to guard a code which needs to be
+// executed when object's ref count becomes zero after release. This method also
+// doesn't need a mutex guard because decrement operation is atomic and only one
+// thread can reach ref count equal to zero, i.e. only a single thread can pass
+// through this check.
+struct ReferenceCounter {
+  ReferenceCounter() : RefCount{1} {}
+
+  // Reset the counter to the initial value.
+  void reset() { RefCount = 1; }
+
+  // Used when retaining an object.
+  void increment() { RefCount++; }
+
+  // Supposed to be used in pi*GetInfo* methods where ref count value is
+  // requested.
+  pi_uint32 load() { return RefCount.load(); }
+
+  // This method allows to guard a code which needs to be executed when object's
+  // ref count becomes zero after release. It is important to notice that only a
+  // single thread can pass through this check. This is true because of several
+  // reasons:
+  //   1. Decrement operation is executed atomically.
+  //   2. It is not allowed to retain an object after its refcount reaches zero.
+  //   3. It is not allowed to release an object more times than the value of
+  //   the ref count.
+  // 2. and 3. basically means that we can't use an object at all as soon as its
+  // refcount reaches zero. Using this check guarantees that code for deleting
+  // an object and releasing its resources is executed once by a single thread
+  // and we don't need to use any mutexes to guard access to this object in the
+  // scope after this check. Of course if we access another objects in this code
+  // (not the one which is being deleted) then access to these objects must be
+  // guarded, for example with a mutex.
+  bool decrementAndTest() { return --RefCount == 0; }
+
+private:
+  std::atomic<pi_uint32> RefCount;
+};
+
+// Base class to store common data
+struct _pi_object {
+  _pi_object() : RefCount{} {}
+
+  // Level Zero doesn't do the reference counting, so we have to do.

```

Comment:
Try to get rid of direct mention of Level Zero in "UR"
Also we need to think how this affects backends that do refcnt, e.g. opencl

Reply:
Done. Yes, we need to see how useful this `_pi_object` is for other adapters.

Also, we probably need to do something with the env variables, like `SYCL_PI_LEVEL_ZERO_SINGLE_THREAD_MODE`. Should we have separate ones for PI and UR?